CorsixTH/Lua/hospitals/player_hospital.lua

```diff
@@ -216,6 +212,143 @@ function PlayerHospital:dailyAdviceChecks()
+    table.insert(msgs, _A.level_progress.three_quarters_lost)
+  elseif rep_lose_ratio > 0.4 or balance_lose_ratio > 0.4 then
+    table.insert(msgs, _A.level_progress.halfway_lost)
+  end
+  return msgs
+end
+
+--! Give advice about general progress to fulfilling all of the goals to win the level
+function PlayerHospital:adviseGoalProgress()
+-- Move to World:World?
+  self.world:determineWinningConditions()
+  local start, goals = self.world.map:getLevelStartState(), self.world.goals
+  if not goals then return end
+
+  local highest_group = 0
+  -- Find highest count group
+  for criteria, v in pairs(goals) do
+    if type(criteria) == "string" then
+      if v["group"] > highest_group then highest_group = v["group"] end
+    end
+  end
+-- End move to World:World?
+-- Combine with World:checkWinningConditions?
+  local criteria_count, win_criteria_met, lose_criteria_met, best_group, best_value = {}, {}, {}
+  -- Collect winning criteria count for each group
+  for g = 1, highest_group do
+    win_criteria_met[g], criteria_count[g] = 0, 0
+    for criteria, t in pairs(goals) do
+      if type(criteria) == "string" and t["group"] == g then
+        -- Win criteria
+        if t["win_value"] then
+          criteria_count[g] = criteria_count[g] + 1
+          if (t["max_min_win"] == 1 and self[criteria] > t["win_value"]) or
+              (t["max_min_win"] == 0 and self[criteria] < t["win_value"]) then
+            win_criteria_met[g] = win_criteria_met[g] + 1
+          end
+        end
+        -- Lose criteria
+        if t["lose_value"] then
+          t["start"] = start[criteria] or 0
+          if (t["max_min_lose"] == 1 and self[criteria] > t["lose_value"]) or
+              (t["max_min_lose"] == 0 and self[criteria] < t["lose_value"]) then
+            lose_criteria_met[criteria] = true -- lose criteria are typically in their own group
+          end
+        end
+      end
+    end
+
+    -- Is this the best performing group in the winning goals?
+    if not best_value or win_criteria_met[g] > best_value then
+      best_value = win_criteria_met[g]
+      best_group = g
+      --[[ If all win criteria in a group are met the level will be won, no advice to give
+      if win_criteria_met[g] > 0 and criteria_count[g] == win_criteria_met[g] then
+        return
+      end]]
+    end
+  end
+
+  local c = 0
+  for _, _ in pairs(lose_criteria_met) do c = c + 1 end
+-- End combine with World:checkWinningConditions
+  local msgs = self:adviseGoalSpecifics(goals)
+
+  -- Generic positive
+  if win_criteria_met[best_group] >= criteria_count[best_group] * 0.75 then
+    table.insert(msgs, _A.level_progress.three_quarters_won)
+  elseif win_criteria_met[best_group] >= criteria_count[best_group] * 0.5 then
+    table.insert(msgs, _A.level_progress.halfway_won)
+  end
+
+  if #msgs == 0 then return end
+for _, m in pairs(msgs) do print(m["text"]) end -- Remove
+  self:giveAdvice(msgs, 1)
+end
+
```

Comment:
Might just be indenting problems here, but would double check something hasn't been left open.